### PR TITLE
Enhanced Encryption

### PR DESCRIPTION
**Pull request by tn3rb**
For the new Square Payment Method @knazart has been working on, we needed a way to be able to encrypt data used for Square using specific standards as well as different keys used during the encryption process.

The existing EE_Encryption class was not capable of this but the changes in this PR are.

To ensure that things are working correctly I've added a few unit tests:

ScreenShot_20210817193640

### CODE REVIEW

**Comment by tn3rb** [comment]
some comments from chat regarding possible usage:

you should be able to use the \EventEspresso\core\services\encryption\openssl\OpenSSLv1 class
(or \EventEspresso\core\services\encryption\openssl\OpenSSLv2 when that is ready) with the Square API.

You can either pass some customized dependencies to it like:

```php
new OpenSSLv1(
    new Base64Encoder(),
    new CipherMethod('aes-128-cbc', 'ee-square-api-cipher-method'),
    new EncryptionKeyManager('ee-square-api-encryption-key-staging', 'ee-square-api-encryption-keys')
);
```

or create a custom class in the Square API that extends OpenSSLv1  and then use that were needed, like:

```php
new SquareEncryption()
```

oh... and to add the second key for production you would have to configure that first before passing the EncryptionKeyManager to OpenSSLv1 something like:

```php
$encryption_key_manager = new EncryptionKeyManager(
    'ee-square-api-encryption-key-staging',
    'ee-square-api-encryption-keys'
);
$encryption_key_manager->addEncryptionKey('ee-square-api-encryption-key-production');

new OpenSSLv1(
    new Base64Encoder(),
    new CipherMethod('aes-128-cbc', 'ee-square-api-cipher-method'),
    $encryption_key_manager
);
```

or again, you could create a custom class that extends EncryptionKeyManager  and configure all of that internally
then just pass an instance of that like:

```php
new OpenSSLv1(
    new Base64Encoder(),
    new CipherMethod('aes-128-cbc', 'ee-square-api-cipher-method'),
    new SquareEncryptionKeyManager()
);
```

if you need to prevent duplicate instances (like a singleton) then you'll need to use the Loader like:

```php
$encryptor = LoaderFactory::getLoader()->getShared(OpenSSLv1::class);
```

but I have yet to register the dependencies for any of the classes,
and you would have to do that for any custom ones, but that's easy (edited) 

oh... and most importantly... the interface for the class you will actually utilize will be like:

```php
$encryptor->encrypt( $text_to_encrypt );
```

and

```php
$encryptor->decrypt( $encrypted_text );
```

but both of those accept a second param for specifying which key to use,
which will use the default specified in the EncryptionKeyManager  constructor
i used 'ee-square-api-encryption-key-staging' in my examples above,
so if you wanted to use the production key, it would be like:

```php
$encryptor->encrypt( $text_to_encrypt, 'ee-square-api-encryption-key-production' );
```

but creating a custom class with constants would be better:

```php
$square_encryptor->encrypt( $text_to_encrypt, SquareEncryption::KEY_PRODUCTION );
```

**Comment by tn3rb** [comment]
will fix lint issues tomorrow

File: core/services/encryption/Base64Encoder.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * encodes  url string with PHP's base64 encoding
+     *
+     * @see http://php.net/manual/en/function.base64-encode.php
+     * @param string $text_string the text to be encoded
+     * @return string
+     */
+    public function encodeUrl($text_string = '')
+    {
+        // you give me nothing??? GET OUT !
```
Comment (manzoorwanijk) [inline core/services/encryption/Base64Encoder.php:93]: 😄

File: core/services/encryption/EncryptionKeyManager.php [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    /**
+     * creates a new encryption key
+     *
+     * @param bool $strong if true (default) will attempt to generate a cryptographically secure key
+     * @return string
+     * @throws Exception
+     */
+    public function generateEncryptionKey($strong = true)
+    {
+        return $strong && PHP_VERSION_ID >= 70100
```
Comment (manzoorwanijk) [inline core/services/encryption/EncryptionKeyManager.php:147]:
```suggestion
     * @param string $type 'STRONG' or 'WEAK'. If 'STRONG' (default) will attempt to generate a cryptographically secure key
     * @return string
     * @throws Exception
     */
    public function generateEncryptionKey($type = 'STRONG')
    {
        return $type === 'STRONG' && PHP_VERSION_ID >= 70100
```

File: core/services/encryption/EncryptionKeyManager.php [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+
+    /**
+     * @param int $bit_depth options are 64, 128, 192, or 256
+     */
+    public function setBitDepth($bit_depth)
+    {
+        $bit_depth       = absint($bit_depth);
+        $this->bit_depth = in_array($bit_depth, $this->bit_depth_options) ? $bit_depth : 128;
```
Comment (manzoorwanijk) [inline core/services/encryption/EncryptionKeyManager.php:201]: Why not strict comparison?
```suggestion
        $this->bit_depth = in_array($bit_depth, $this->bit_depth_options, true) ? $bit_depth : 128;
```

File: core/services/encryption/openssl/CipherMethod.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            // then grab the first item from the list (we'll add it back later if it is good)
+            $cipher_method_to_test = array_shift($this->cipher_methods);
+            if ($cipher_method_to_test === null) {
+                throw new RuntimeException(
+                    esc_html__(
+                        'OpenSSL support appears to be enabled on the server, but no cipher methods are available. Please contact the server administrator.',
+                        'event_espresso'
+                    )
+                );
+            }
```
Comment (manzoorwanijk) [inline core/services/encryption/openssl/CipherMethod.php:152]: 👍

File: core/services/encryption/openssl/CipherMethod.php [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    /**
+     * @param string $cipher_method
+     * @param false  $throw_exception
+     * @return bool
+     */
+    protected function validateCipherMethod($cipher_method, $throw_exception = false)
+    {
+        // verify that the requested cipher method is actually installed and can produce an initialization vector
+        if (
+            in_array($cipher_method, $this->installed_cipher_methods)
```
Comment (manzoorwanijk) [inline core/services/encryption/openssl/CipherMethod.php:187]:
```suggestion
            in_array($cipher_method, $this->installed_cipher_methods, true)
```

File: tests/mocks/core/services/encryption/EncryptionKeyManagerMock.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * saves encryption keys from db
+     *
+     * @return bool
+     */
+    public function saveEncryptionKeys($initialize = false)
+    {
+        return parent::saveEncryptionKeys($initialize);
+    }
+}
```
Comment (manzoorwanijk) [inline tests/mocks/core/services/encryption/EncryptionKeyManagerMock.php:56]: EOL

File: tests/mocks/core/services/encryption/OpenSSLv1Mock.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace EventEspresso\tests\mocks\core\services\encryption;
+
+use EventEspresso\core\services\encryption\openssl\OpenSSLv1;
+
+class OpenSSLv1Mock extends OpenSSLv1
+{
+
+}
```
Comment (manzoorwanijk) [inline tests/mocks/core/services/encryption/OpenSSLv1Mock.php:10]: EOL

File: tests/mocks/core/services/encryption/OpenSSLv2Mock.php [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace EventEspresso\tests\mocks\core\services\encryption;
+
+use EventEspresso\core\services\encryption\openssl\OpenSSLv2;
+
+class OpenSSLv2Mock extends OpenSSLv2
+{
+
+}
```
Comment (manzoorwanijk) [inline tests/mocks/core/services/encryption/OpenSSLv2Mock.php:10]: EOL

File: tests/testcases/core/services/encryption/EncryptionKeyManagerTest.php [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+            [7, 8],
+            [8, 8],
+            [16, 16],
+            [100, 100],
+            [128, 128],
+            [256, 128],
+            ['abc', 8],
+            ['%$@#', 8],
+            [null, 8],
+            [true, 8],
```
Comment (manzoorwanijk) [inline tests/testcases/core/services/encryption/EncryptionKeyManagerTest.php:146]: 👍

**Review by manzoorwanijk — APPROVED** [review]
LGTM 👍

File: core/services/encryption/EncryptionKeyManager.php (hunk 2 above)
Comment (tn3rb) [reply core/services/encryption/EncryptionKeyManager.php:147]: ??? 

is this for readability on the function calling side of things?

I get that argument, but prefer having PhpStorm tell me what those bools mean:

ScreenShot_20210818104315

File: core/services/encryption/EncryptionKeyManager.php (hunk 3 above)
Comment (tn3rb) [reply core/services/encryption/EncryptionKeyManager.php:201]: oh good catch, thnx, i totally missed that, will fix

File: core/services/encryption/openssl/CipherMethod.php (hunk 5 above)
Comment (tn3rb) [reply core/services/encryption/openssl/CipherMethod.php:187]: thnx, will fix

File: tests/mocks/core/services/encryption/EncryptionKeyManagerMock.php (hunk 6 above)
Comment (tn3rb) [reply tests/mocks/core/services/encryption/EncryptionKeyManagerMock.php:56]: thnx will fix

File: core/services/encryption/EncryptionKeyManager.php (hunk 2 above)
Comment (manzoorwanijk) [reply core/services/encryption/EncryptionKeyManager.php:147]: Yes it's for readability on the function calling side of things.

> I get that argument, but prefer having PhpStorm tell me what those bools mean

Cool

**Review by knazart — APPROVED** [review]
👌

**Pull request merged**